### PR TITLE
Implement observation removal

### PR DESCRIPTION
**Pull request by Lycopersicum**

### CODE REVIEW

File: examples/rest-server/rest-subscriptions.c [hunk 1]
```diff
@@ -182,3 +190,104 @@ int rest_subscriptions_put_cb(const ulfius_req_t *req, ulfius_resp_t *resp, void
...
+        ulfius_set_empty_body_response(resp, 404);
+        return U_CALLBACK_CONTINUE;
+    }
+
+    res = lwm2m_observe_cancel(
+            rest->lwm2m, client->internalID, &uri,
+            rest_unobserve_cb, observe_context
+    );
+
+    if (res != 0)
```
Comment (GiedriusM) [inline examples/rest-server/rest-subscriptions.c]: Do not return if `res == COAP_404_NOT_FOUND`, just continue with removal and response, but print a warning about mismatch between rest and lwm2m.

File: examples/rest-server/rest-subscriptions.c [hunk 2]
```diff
@@ -182,3 +190,104 @@ int rest_subscriptions_put_cb(const ulfius_req_t *req, ulfius_resp_t *resp, void
...
+    {
+        return err;
+    }
+
+    rest_list_remove(rest->observeList, observe_context->response);
+
```
Comment (GiedriusM) [inline examples/rest-server/rest-subscriptions.c]: Free memory allocated in observe registration (PUT).

File: examples/rest-server/rest-subscriptions.c [hunk 3]
```diff
@@ -182,3 +190,104 @@ int rest_subscriptions_put_cb(const ulfius_req_t *req, ulfius_resp_t *resp, void
...
+    jresponse = json_object();
+    json_object_set_new(jresponse, "async-response-id", json_string(observe_context->response->id));
```
Comment (GiedriusM) [inline examples/rest-server/rest-subscriptions.c]: Status code 204 should be returned with no payload.

File: examples/rest-server/rest-subscriptions.c [hunk 4]
```diff
@@ -57,11 +57,9 @@ static void rest_observe_cb(uint16_t clientID, lwm2m_uri_t *uriP, int count, lwm
 }
 
 static void rest_unobserve_cb(uint16_t clientID, lwm2m_uri_t *uriP, int count, lwm2m_media_type_t format,
-                              uint8_t *data, int dataLength, void *context)
```
Comment (GiedriusM) [inline examples/rest-server/rest-subscriptions.c]: Prieš tai buvo teisingas alignmentas

File: examples/rest-server/rest-subscriptions.c [hunk 5]
```diff
@@ -182,3 +188,114 @@ int rest_subscriptions_put_cb(const ulfius_req_t *req, ulfius_resp_t *resp, void
...
+    );
+
+    if (res == COAP_404_NOT_FOUND)
+    {
+        fprintf(stdout, "[WARNING] LwM2M and restserver subscriptions mismatch!");
+    } else if (res != 0) {
+        return err;
+    }
+
+    fprintf(stdout, "[UNOBSERVE-RESPONSE] id=%s\n", observe_context->response->id);
```
Comment (valinskas) [inline examples/rest-server/rest-subscriptions.c]: stdout -> stderr, just a sanity

File: examples/rest-server/rest-subscriptions.c [hunk 6]
```diff
@@ -182,3 +188,114 @@ int rest_subscriptions_put_cb(const ulfius_req_t *req, ulfius_resp_t *resp, void
...
+    len = snprintf(path, sizeof(path), "/subscriptions/%s/", name);
+
+    if (req->http_url == NULL || strlen(req->http_url) >= sizeof(path) || len >= sizeof(path))
+    {
+        fprintf(stderr, "%s(): invalid http request (%s)!\n", __func__, req->http_url);
+        return U_CALLBACK_ERROR;
+    }
+
+    // this is probaly redundant if there's only one matching ulfius filter
+    if (strncmp(path, req->http_url, len) != 0)
```
Comment (valinskas) [inline examples/rest-server/rest-subscriptions.c]: strncmp(,,len) will compare only 'len' bytes. Is that ok ? Consider this for example:

path = /subscriptions/some/
req->http_url = /subscriptions/some//x/y/z

File: examples/rest-server/rest-subscriptions.c (hunk 5 above)
Comment (GiedriusM) [reply examples/rest-server/rest-subscriptions.c]: Kodėl stderr? Čia ne klaida, o debugas.

File: examples/rest-server/rest-subscriptions.c (hunk 6 above)
Comment (GiedriusM) [reply examples/rest-server/rest-subscriptions.c]: That's exactly what's done here:
path = /subscriptions/some/
req->http_url = /subscriptions/some//x/y/z

should pass and the next section should check /x/y/z part

Comment (valinskas) [reply examples/rest-server/rest-subscriptions.c]: ok

**Pull request merged**